runtime/nativeMain/src/kotlinx/benchmark/NativeBlackhole.kt

```diff
@@ -1,35 +1,61 @@
 package kotlinx.benchmark
 
-import kotlinx.cinterop.pin
+import kotlinx.cinterop.convert
+import kotlinx.cinterop.toByte
+import kotlin.concurrent.Volatile
+import kotlin.native.ref.WeakReference
+import kotlin.random.Random
+import kotlinx.benchmark.impl.*
 
+@OptIn(ExperimentalStdlibApi::class)
 actual class Blackhole {
+    /*
+     * For primitive types, DCE may be omitted by creating an artificial use via inline asm.
+     * For objects, there is no cheap way to obtain something that could be passed to a native function with such
+     * inline asm, so instead we're generating a sequence of random numbers and as soon as we get zero (that's highly
+     * unlikely to happen) an object will be wrapped into a weak reference and stored into a field.
+     */
+    @Volatile
+    private var lcg: Int = Random.nextInt()
+    @Volatile
+    internal var ref: WeakReference<Any>? = null
+
     actual fun consume(obj: Any?) {
-        obj?.pin()
+        // https://en.wikipedia.org/wiki/Linear_congruential_generator
+        val next = lcg * 1664525 + 1013904223
+        if (next == 0) {
+            if (obj !== null) {
+                ref = WeakReference(obj)
+            }
+        }
+        lcg = next
```

Comment:
I'm a bit concerned it is a full memory (seq consistent) barriered write on each `consume` that potentially can affect measurement precision.

Could you please elaborate a bit on why the conditional write is not enough, taking into account the fact that `lcg` already is `volatile`?

Reply:
The store should be eliminated, it's just a mistake. I'll rewrite this code.